README.md
# Rendition

[![npm version](https://badge.fury.io/js/rendition.svg)](http://badge.fury.io/js/rendition)
[![Dependency Status](https://img.shields.io/david/balena-io-modules/rendition.svg)](https://david-dm.org/balena-io-modules/rendition)
[![style: styled-components](https://img.shields.io/badge/style-%F0%9F%92%85%20styled--components-orange.svg?colorB=daa357&colorA=db748e)](https://github.com/styled-components/styled-components)

A library of UI components, built using [React][react], [recompose][recompose], [styled-components][styled-components] and [styled-system][styled-system].

## Table of Contents

* [Installation](#installation)
* [Usage](#usage)
* [Components](#components)
* [Development](#development)
* [Testing](#testing)
* [Upgrading](#testing)

## Installation

```
npm install --save rendition
```

## Usage

Wrap your application in the `Provider` component and start using components!

```jsx
import React from 'react'
import ReactDOM from 'react-dom'
import { Button, Box, Provider } from 'rendition';

ReactDOM.render(
  <Provider>
    <Box my={3} mx={['auto', 15]}>
      <Button primary emphasized>Click me</Button>
    </Box>
  </Provider>,
  document.getElementById('root')
);
```

## Components

For an interactive demo of all components, see https://balena-io-modules.github.io/rendition

- [Accordion](#accordion)
- [Alert](#alert)
- [ArcSlider](#arcslider)
- [Badge](#badge)
- [Banner](#banner)
- [Box](#box)
- [Button](#button)
- [ButtonGroup](#buttongroup)
- [Card](#card)
- [Checkbox](#checkbox)
- [Container](#container)
- [Copy](#copy)
- [Divider](#divider)
- [DropDownButton](#dropdownbutton)
- [Filters](#filters)
- [Fixed](#fixed)
- [Flex](#flex)
- [Heading](#heading)
- [HighlightedName](#highlightedname)
- [Img](#img)
- [Input](#input)
- [Link](#link)
- [List](#list)
- [Modal](#modal)
- [Navbar](#navbar)
- [Notifications](#notifications)
- [Pager](#pager)
- [ProgressBar](#progressbar)
- [RadioButton](#radiobutton)
- [RadioButtonGroup](#radiobuttongroup)
- [Search](#search)
- [Select](#select)
- [Spinner](#spinner)
- [Steps](#steps)
- [Table](#table)
- [Tabs](#tabs)
- [Tag](#tag)
- [Terminal](#terminal)
- [TextWithCopy](#textwithcopy)
- [Textarea](#textarea)
- [Tooltips](#tooltips)
- [Txt](#txt)
- [Markdown](#markdown)
- [MarkdownEditor](#markdowneditor)
- [Mermaid](#mermaid)
- [Form](#form)

### Accordion

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Accordion/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `items`     | `[{label: string \| node; panel: string \| node }]` | - | true | renders label as title and panel as body of accordion     |

### Alert

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Alert/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `primary`     | `boolean` | - | - | If true, use the `primary` theme color     |
| `secondary`   | `boolean` | - | - | If true, use the `secondary` theme color   |
| `tertiary`    | `boolean` | - | - | If true, use the `tertiary` theme color    |
| `quarternary` | `boolean` | - | - | If true, use the `quarternary` theme color |
| `danger`      | `boolean` | - | - | If true, use the `danger` theme color      |
| `warning`     | `boolean` | - | - | If true, use the `warning` theme color     |
| `success`     | `boolean` | - | - | If true, use the `success` theme color     |
| `info`        | `boolean` | - | - | If true, use the `info` theme color        |
| `emphasized`  | `boolean` | - | - | If true, use the use bolder colors and a larger size |
| `plaintext` | `boolean` | - | - | If true, the alert will be rendered without a border or a background |
| `prefix` | <code>JSX.Element &#124; string &#124; false</code> | - | - | Set a prefix on the alert message, if this prop is set to false, the default prefix will not be shown |
| `onDismiss` | `() => void` | - | - | A function that is called when dismissing an alert

### ArcSlider

A slider input that is displayed as an arc. This component will scale in size to
fit it's container. A label can be added by placing a child element inside this
component.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/ArcSlider/story.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                                   |
| --------------|-----------|-----------|------------|-------------------------------------------------------------- |
| `onValueChange`    | `(value: number) => void` | - | - | A function that is called when the slider value changes, this will always be a value between 0 and 1 |
| `value`       | `number`  | -         | -          | A number between 0 and 1 that represents the progress         |
| `fillColor`   | `string`  | -         | -          | A CSS color string to use for the color of the "filled" part of the arc |
| `background`  | `string`  | -         | -          | A CSS color string to use for the color of the arc track      |

#### Inheritance

The properties of the [`Box`](#box) component are also available.
### Badge

Badges should be used to present data that is not generated by the user (otherwise, tags should be used).
The badge can use any of the pre-defined shades, from which the background and text color are generated.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Badge/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `shade`        | `number` | - | - | The index for one of the pre-determined shades for the badge |
| `children`  | `string` | - | ✓ | The text to display inside the badge |

### Banner

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Banner/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `backgroundImage`| `string`| - | - | The path to an image that should be displayed in the background |
| `minHeight`| `string`| - | - | The minimum height of the `Banner`, eg `180px` |


### Box

Displays a block level element.

The basic building block of a rendition application.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Grid/Box.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `flex`      | <code>string &#124; string[]</code> | - | - | Sets `flex`, if the value is an array, sets a responsive style corresponding to the theme's breakpoints
| `order`      | <code>number &#124; string &#124; Array&lt;number &#124; string&gt;</code> | - | - | Sets `order`, if the value is an array, sets a responsive style corresponding to the theme's breakpoints


### Button

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Button/story.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `primary`     | `boolean` | -         | -          | If true, use the `primary` theme color               |
| `secondary`   | `boolean` | -         | -          | If true, use the `secondary` theme color             |
| `tertiary`    | `boolean` | -         | -          | If true, use the `tertiary` theme color              |
| `quarternary` | `boolean` | -         | -          | If true, use the `quarternary` theme color           |
| `danger`      | `boolean` | -         | -          | If true, use the `danger` theme color                |
| `warning`     | `boolean` | -         | -          | If true, use the `warning` theme color               |
| `success`     | `boolean` | -         | -          | If true, use the `success` theme color               |
| `info`        | `boolean` | -         | -          | If true, use the `info` theme color                  |
| `light`       | `boolean` | -         | -          | If true, use white background and default text color |
| `disabled`    | `boolean` | -         | -          | If true, disabled the button |
| `outline`     | `boolean` | -         | -          | If true, the button will have a transparent background, and the border and text color will match |
| `plain`       | `boolean` | -         | -          | If true, render the button without padding, background or border |
| `underline`   | `boolean` | -         | -          | Similar to the `plaintext` prop, but displays a line underneath the button text |
| `icon`        | `JSX.Element` | -         | -          | Optionally provide a JSX element that will be rendered before the text inside the button |
| `compact`        | `boolean[]` | [false]   | -          | Optionally renders the label according to the value inside the array for each breakpoint |

### ButtonGroup

Wrapper for buttons to make them stick together.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/ButtonGroup/story.js)
### Card

Section containing content and actions on the same topic. 

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Card/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| -------|--------|-----------|------------|-------------- |
| `title`  | `string` | - | - | The title of the card |
| `cta`  | `JSX.Element` | - | - | React component added to the header |
| `rows`  | `JSX.Element[]` | - | - | Subsections separated by a horizontal separator |
| `minHeight`  | `string` | `auto` | - | CSS minHeight property |
| `small`  | `boolean` | - | - | If true, reduces the scale of the card |
| `children`  | `JSX.Element` | - | - | Any content that is internally wrapped in a Box |

### Checkbox

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Checkbox/story.js)

#### Props

`Checkbox` implements all properties from [Grommet's CheckBox component](https://v2.grommet.io/checkbox)

### Container

A padded container with a responsive width.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Container/story.js)


### Copy

Wrapper that adds a "copy to clipboard" button to any component and copies the passed content to the user's clipboard.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Copy/story.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `content`      | `string` | -         | ✓          | The value that should be copied to the clipboard
| `show`      | <code>'hover' &#124; 'always'</code> | - | - | Optionally show the copy button on hover or always show the button
| `onClick`      | `(content: string) => void` | - | - | onClick handler, useful if you wish to do other actions after content was copied
| `children`      | `React.ReactNode` | -         | - | The content next to which the clipboard button should be shown

### Divider

A styled horizontal rule.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Divider/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| -------|--------|-----------|------------|-------------- |
| `height`  | `number` | `0.5` | - | The height of the divider |
| `color`  | `string` | `quartenary.main` | - | The color of the divider |
| `type`  | `'solid' | 'dashed'` | `'solid` | - | The type of divider |
| `children`  | `string` | - | - | Textual explanation to go in the middle of the divider |


### DropDownButton

Displays a button with an attached dropdown list, `children` of the component
are rendered inside a dropdown list.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/DropDownButton/story.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------------- | ------ | --------- | ---------- | ------------- |
| `primary`     | `boolean` | -         | -          | If true, use the `primary` theme color               |
| `secondary`   | `boolean` | -         | -          | If true, use the `secondary` theme color             |
| `tertiary`    | `boolean` | -         | -          | If true, use the `tertiary` theme color              |
| `quarternary` | `boolean` | -         | -          | If true, use the `quarternary` theme color           |
| `danger`      | `boolean` | -         | -          | If true, use the `danger` theme color                |
| `warning`     | `boolean` | -         | -          | If true, use the `warning` theme color               |
| `success`     | `boolean` | -         | -          | If true, use the `success` theme color               |
| `info`        | `boolean` | -         | -          | If true, use the `info` theme color                  |
| `emphasized`  | `boolean` | -         | -          | If true, use a larger size                           |
| `square`      | `boolean` | -         | -          | If true, render the button with equal length width and height |
| `disabled`      | `boolean` | -         | -          | If true, disabled the button |
| `label`      | `JSX.Element` | -         | -          | Optionally provide a JSX element that will be displayed inside the main button |
| `border`      | `boolean` | -         | -          | If true, place a border between each item in the dropdown |
| `joined`      | `boolean` | -         | -          | If true, render the component as a single button instead of two |
| `noListFormat`      | `boolean` | -         | -          | If true, render children as a single JSX element instead of iterating over each of them |
| `alignRight`      | `boolean` | -         | -          | If true, put the dropdown list on the right  |
| `listMaxHeight`      | `string \| number` | -         | -          | If setted, it defines the maximum height of the dropdown list |
| `compact`        | `boolean[]` | [false]    | -          | Optionally renders the label according to the value inside the array for each breakpoint |

### Filters

A component that can be used for generating filters in the form of [json schema](http://json-schema.org/) objects and saving sets of filters as "views".
The filters created by this component can be used to filter a collection of
objects using the `SchemaSieve` object.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Filters/story.js)

#### Schema

The `Filters` component requires a `schema` property which should be a json
schema that defines the shape of the objects you want to filter. For example if
you want to filter on a collection that looks like this:

```
[
  {
    name: 'Bulbasaur',
    caught: true,
  },
  {
    name: 'Pikachu',
    caught: true,
  },
  {
    name: 'Dratini',
    caught: false,
  }
]
```

You would define a schema that looks like this:

```
{
  type: 'object',
  properties: {
    name: {
      title: 'Name',
      type: 'string'
    },
    caught: {
      title: 'Has been caught',
      type: 'boolean'
    }
  }
}
```

If you provide a `title` property, it will be used to label the field when
filtering, otherwise the field name will be used.

##### Views

Views represent a set of filters, along with an id and a name. This is a useful
feature for storing a set filters and loading it again at a later point.
A view can optionally have a `scope` property, which will correspond to the
`slug` of a view scope, if you have provided one in the `Filters` property
`viewScopes` property. Scopes allow you to easily add an extra layer of
granularity/grouping to views that are generated. If you provide view scopes,
the user can select a scope when creating a new view.

A view scope has the following properties:

| Name          | Type      | Description                                          |
| ------------- | --------- | ---------------------------------------------------- |
| slug            | `string`  | A unique identifier for the scope                  |
| name          | `string`  | A descriptive name for the scope                     |
| label       | `string`  | An optional label to use for this scope when creating a view  |

A view has the following properties:

| Name          | Type      | Description                                          |
| ------------- | --------- | ---------------------------------------------------- |
| id            | `string`  | A unique identifier for the view                     |
| name          | `string`  | A descriptive name for the view                      |
| filters       | `string`  | An array of json schemas                             |
| scope       | <code>string &#124; null</code>  | The slug of a view scope, or `null` if now scopes are provided |

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `schema`    | `object` | - | ✓ | A json schema describing the shape of the objects you want to filter |
| `disabled`    | `boolean` | -         | -          | If true, disable the entire `Filters` interface
| `filters`    | `object[]` | -         | -          | An array of json schemas to be displayed as the currently selected filters, typically used when loading when loading filters from storage |
| `views`    | `object[]` | -         | -          | An array of views, as described above, typically used when loading when loading views from storage |
| `viewScopes`    | `object[]` | - | - | An array of view scopes, as described above |
| `onFiltersUpdate`    | `(filters: object[]) => void` | - | - | A function that is called when filters are updated |
| `onViewsUpdate`    | `(views: object[]) => void` | - | - | A function that is called when views are updated |
| `addFilterButtonProps` | `object` | - | - | Properties that are passed to the "Add filter" button, these are the same props used for the [`Button`](#button) component |
| `viewsMenuButtonProps` | `object` | - | - | Properties that are passed to the "Views" button, these are the same props used for the [`DropDownButton`](#dropdownbutton) component |
| `renderMode` | <code>string &#124; string[]</code> | - | - | Controls which parts of the `Filters` interface are displayed. One of `all`, `add`, `search`, `views`, `summary`, or an array containing any of these values |
| `dark`    | `boolean` | -         | -          | If true, Set the `Filters` component against a dark background |
| `compact`    | `boolean[]` | -         | -          | Accept a boolean for each rendition breakpoint. If true remove `Filters` labels  |
### Fixed

Displays an element with a [`fixed`](https://developer.mozilla.org/en-US/docs/Web/CSS/position#fixed) position.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Fixed/story.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `top`      | <code>boolean &#124; ResponsiveStyle</code> | - | - | Sets the distance to the top of the containing block. If true, sets the value to zero
| `right`      | <code>boolean &#124; ResponsiveStyle</code> | - | - | Sets the distance to the right of the containing block. If true, sets the value to zero
| `bottom`      | <code>boolean &#124; ResponsiveStyle</code> | - | - | Sets the distance to the bottom of the containing block. If true, sets the value to zero
| `left`      | <code>boolean &#124; ResponsiveStyle</code> | - | - | Sets the distance to the left of the containing block. If true, sets the value to zero
| `z`      | `ResponsiveStyle` | - | - | Sets the z-index of the component |
### Flex

Displays an element using [flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox).

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Grid/Flex.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `alignItems`      | <code>string &#124; string[]</code> | - | - | Sets `align-items`, if the value is an array, sets a responsive style corresponding to the theme's breakpoints
| `justifyContent`      | <code>string &#124; string[]</code> | - | - | Sets `justify-content`, if the value is an array, sets a responsive style corresponding to the theme's breakpoints
| `flexDirection`      | <code>string &#124; string[]</code> | - | - | Sets `flex-direction`, if the value is an array, sets a responsive style corresponding to the theme's breakpoints
| `flexWrap`      | <code>string &#124; string[]</code> | - | - | Sets `flex-wrap` css property
### Heading

A component that displays a heading. By default an `<h3>` tag is used. The exact
heading type can be specifed by appending the element name to the component, for
example `<Heading.h1>`, `<Heading.h2>`, `<Heading.h3>` etc.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Heading/story.js)

### HighlightedName

By default, the background color of a `HighlightedName` component is generated
automatically from its `children` (which must be a string), though this can be overridden.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/HighlightedName/story.js)

#### Props

| Name       | Type     | Default                             | Required | Description                                         |
|------------|----------|-------------------------------------|----------|-----------------------------------------------------|
| `children` | `string` | -                                   | ✓        | The text to display inside the highlighted name     |
| `bg`       | `string` | autogenerated from `children`       | -        | If passed, it will be used as the background color of the component |
| `color`    | `string` | `theme.colors.text.main` or `white` | -        | If passed, it will be used as the text color of the component |

### Img

Displays an image.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Img/story.js)

#### Inheritance

The attributes of an [`<img>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img) element are also available.

### Input

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Input/story.js)

#### Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `emphasized`  | `boolean` | -         | -          | If true, use a larger size                           |
| `monospace`  | `boolean` | - | - | If true, render text in a monospace font |
| `onChange`  | `(e: Event) => void` | - | - | A function that is called when the input value changes |

#### Inheritance

The attributes of an [`<input>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) element are also available.



### Link

Displays an anchor link.

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/components/Link/story.js)

#### Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `blank`  | `boolean` | - | - | If true, open the link in a new tab |
| `disabled`  | `boolean` | - | - | If true, disable the link |

#### Inheritance


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v14.12.0
## (2020-05-27)

* update tab title to keep font colors [StefKors]
* generate screenshots [StefKors]
* Add support for scroll snapping [StefKors]
* Add support for scrollable tabs [Graham McCulloch]

# v14.11.7
## (2020-05-21)

* Update react-simplemde-editor to 4.1.1 [StefKors]

# v14.11.6
## (2020-05-21)

* Bump node engine to support grommet@2.13.0 [StefKors]
* Update Screenshots [StefKors]
* Bump grommet and fix typings [Stevche Radevski]

# v14.11.5
## (2020-05-11)

* Update to prettier 2.x to match balena-lint [Pagan Gazzard]
* Update to balena-lint 5.x [Pagan Gazzard]

# v14.11.4
## (2020-05-11)

* CI: Remove nodejs version restriction now the minimum is 8 [Pagan Gazzard]

# v14.11.3
## (2020-05-07)

* Switch to optional chaining instead of lodash/get where possible [Pagan Gazzard]
* Switch to native object type checking [Pagan Gazzard]
* Switch to native boolean type checking [Pagan Gazzard]
* Switch to native function type checking [Pagan Gazzard]
* Switch to native array type checking [Pagan Gazzard]
* Switch to native string type check [Pagan Gazzard]
* Switch to native Object.assign [Pagan Gazzard]

# v14.11.2
## (2020-04-30)

* fix: internal nullable prop [JSReds]

# v14.11.1
## (2020-04-29)

* fix: overflow bottom padding [JSReds]

# v14.11.0
## (2020-04-28)

* feat: Add sanitizer options to Markdown component [Graham McCulloch]

# v14.10.0
## (2020-04-28)

* feat: Add support for ordered "Steps" [Graham McCulloch]

# v14.9.2
## (2020-04-21)

* Theme: Allow setting string values to the input & button weight [Thodoris Greasidis]
* Provider: Type the theme property as deep partial [Thodoris Greasidis]

# v14.9.1
## (2020-04-21)

* Make value and onChange optional for Search to allow uncontrolled [Stevche Radevski]

# v14.9.0
## (2020-04-17)

* Apply text color from theme globally [Stevche Radevski]

# v14.8.0
## (2020-04-15)

* Make the theme object the source of the exported Theme interface [Thodoris Greasidis]

# v14.7.0
## (2020-04-11)

* Implement Accordion component [Amit Solanki]

# v14.6.0
## (2020-04-02)

* add: compact prop on Button and Dropdown [JSReds]
* revert: previous filters refactor [JSReds]

# v14.5.0
## (2020-04-02)

* MarkdownEditor: Update tsifyied styles of easymde to v2.10.0 [Thodoris Greasidis]
* Select: Update snap based on new Grommet [Thodoris Greasidis]

# v14.4.4
## (2020-04-01)

* refactor: schema sieve service function [JSReds]

# v14.4.3
## (2020-04-01)

* fix: filters search bar [JSReds]

# v14.4.2
## (2020-03-30)

* Make spinner use tertiary color [Stevche Radevski]

# v14.4.1
## (2020-03-30)

* Bump grommet to 2.11.3 [Stevche Radevski]

# v14.4.0
## (2020-03-29)

* Add: Compact prop on filters [JSReds]
* Refactor: Filters component [JSReds]

# v14.3.3
## (2020-03-27)

* Update the spinner typings to accept a custom label component [Stevche Radevski]

# v14.3.2
## (2020-03-27)

* Fix radio button not displayed correctly [Stevche Radevski]

# v14.3.1
## (2020-03-26)

* Remove onClick event on DropDownButton toggle button [Lorenzo Alberto Maria Ambrosi]

# v14.3.0
## (2020-03-24)

* Support schema.examples in Form [Stevche Radevski]

# v14.2.2
## (2020-03-20)

* Provider: Fix the theme prop typings [Thodoris Greasidis]

## 14.2.1 - 2020-03-18

* Fix Dropdown Chevron Icon Alignment [Amit Solanki]

## 14.2.0 - 2020-03-16

* Add onPageChange callback to table component [StefKors]

## 14.1.0 - 2020-03-16

* Lock Grommet version to 2.11.1. [Amit Solanki]
* Update Alert design, improving accessibility [Amit Solanki]

## 14.0.0 - 2020-03-12

* Update Grommet, Styled components and supporting libs [Amit Solanki]

## 13.4.0 - 2020-03-11

* Add a spinner component [Stevche Radevski]

## 13.3.3 - 2020-03-04

* Aria-label on alert close button [Amit Solanki]

## 13.3.2 - 2020-03-03

* Add: useBreakpoint hook to provider [JSReds]

## 13.3.1 - 2020-02-28

* Import react-notifications-component theme with styled-components [Lorenzo Alberto Maria Ambrosi]
* Replace animate.css with keyframes and classes [Lorenzo Alberto Maria Ambrosi]

## 13.3.0 - 2020-02-27

* Allow inlining objects in Form [Stevche Radevski]
* Fix: in a diffrent way [JSReds]
* Fix: markdown test [JSReds]
* Fix: build and bump typescript version [JSReds]

## 13.2.3 - 2020-02-17

* Fix Tooltip display for badges [Amit Solanki]

## 13.2.2 - 2020-02-14

* Update background and text shades for badges [Stevche Radevski]

## 13.2.1 - 2020-02-14

* Increase font size for HighlightedName component [Stevche Radevski]

## 13.2.0 - 2020-02-12

* Fix: modal enter key check [JSReds]

## 13.1.0 - 2020-02-10

* Add new shades for badge and adjust some existing ones [Stevche Radevski]
* Add new shades for badge and adjust some existing ones [Stevche Radevski]

## 13.0.4 - 2020-02-10

* Use resin-lint 3.x for auto-fixing and most of the linting rules [Pagan Gazzard]

## 13.0.3 - 2020-02-10

* Allow undefined to be returned from render functions [Pagan Gazzard]

## 13.0.2 - 2020-02-10

* Switch xterm.js to dynamic char atlas for performance [Pagan Gazzard]
* Fix filter using is/is_not operators for objects [Amit Solanki]

## 13.0.1 - 2020-02-04

* Update: docs [Andrea Rosci]

## 13.0.0 - 2020-01-22

* Styled system: Add styled system props [JSReds]

## 12.0.1 - 2020-01-16

* Refactor utils file and remove unnecessary functions [Stevche Radevski]

## 12.0.0 - 2020-01-16

* Update badge to match the new styleguide [Stevche Radevski]
* Remove filter migrations [Stevche Radevski]

## 11.36.0 - 2020-01-16

* Minor improvements and refactoring for the HighlightedName component [Stevche Radevski]
* Feat(HighlightedName): Add component [Banana]

## 11.35.7 - 2020-01-15

* Fix dependency regression for storybook and remove unnecessary Providers [Stevche Radevski]

## 11.35.6 - 2020-01-09


src/theme.ts
import { px } from 'styled-system';
import get from 'lodash/get';
import { Omit } from './common-types';

const primary = '#00AEEF';
const secondary = '#2A506F';
const tertiary = '#527699';
const quartenary = '#DDE1f0';

const danger = '#FF423D';
const warning = '#FCA321';
const success = '#1AC135';
const info = '#1496E1';

interface ColorVariants {
	main: string;
	light?: string;
	dark?: string;
	semilight?: string;
}

const colors = {
	primary: {
		main: primary,
		semilight: '#aedff9',
		light: '#08bcff',
		dark: '#009dd7',
	},
	secondary: {
		main: secondary,
		semilight: '#abb9c5',
		light: '#2e587a',
		dark: '#23445e',
	},
	tertiary: {
		main: tertiary,
		light: '#5b82a7',
		semilight: '#bbc8d6',
		dark: '#456482',
	},
	quartenary: {
		main: quartenary,
		light: '#f8f9fd',
		dark: '#b7bed3',
	},

	danger: {
		main: danger,
		semilight: '#ffa1a1',
		light: '#feebeb',
		dark: '#eb0800',
	},
	warning: {
		main: warning,
		semilight: '#fdd190',
		light: '#fef3e5',
		dark: '#ad6800',
	},
	success: {
		main: success,
		semilight: '#8ce09a',
		light: '#e8f8ea',
		dark: '#138b27',
	},
	info: {
		main: info,
		semilight: '#90cbee',
		light: '#e8f5fc',
		dark: '#107dbc',
	},

	text: {
		main: '#2A506F',
		light: '#527699',
		dark: '#23445e',
	},

	statusIdle: {
		main: '#89c683',
	},
	statusConfiguring: {
		main: '#ffb25e',
	},
	statusUpdating: {
		main: '#75C5F5',
	},
	statusPostProvisioning: {
		main: '#aa96d5',
	},
	statusOffline: {
		main: '#fd7c7c',
	},
	statusInactive: {
		main: '#d3d6db',
	},

	gray: {
		main: '#c6c8c9',
		light: '#f4f4f4',
		dark: '#9f9f9f',
	},
};

const defaultControlHeight = 38;
const emphasizedControlHeight = 48;

export const breakpoints = [576, 768, 992, 1200];

export const space = [
	0,
	4,
	8,
	16,
	defaultControlHeight,
	emphasizedControlHeight,
	128,
];

export const fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 72, 96];
export const weights = [400, 700];

export const radius = 3;

export const font = `'Nunito', Arial, sans-serif`;
export const titleFont = `CircularStd, Arial, sans-serif`;
export const monospace = `'Ubuntu Mono', 'Courier New', monospace`;
export const lineHeight = 1.5;

const theme = {
	breakpoints,
	space,
	fontSizes,
	weights,
	font,
	titleFont,
	monospace,
	lineHeight,
	colors,
	radius,
	global: {
		font: {
			family: font,
			size: px(fontSizes[1]),
			height: lineHeight,
		},
		control: {
			disabled: {
				opacity: 0.4,
			},
			border: {
				radius: '4px',
				color: colors.quartenary.main,
			},
		},
		colors: {
			focus: 'none',
			placeholder: colors.secondary.semilight,
		},
		selected: {
			background: colors.primary.main,
		},
		hover: {
			background: {
				color: colors.quartenary.main,
				opacity: 1,
			},

			color: {
				dark: 'inherit',
				light: 'inherit',
			},
		},
		active: {
			background: {
				color: colors.quartenary.main,
				opacity: 1,
			},

			color: {
				dark: 'inherit',
				light: 'inherit',
			},
		},
		drop: {
			border: {
				radius: '4px',
			},
			zIndex: 45,
			// TODO: Border color should be tertiary.main once the TODOs in the Select component are resolved.
			extend: `
				color: ${colors.secondary.main}; border: 1px solid ${colors.quartenary.main};
				animation-duration: 0s;
			`,
		},
		input: {
			weight: 400 as string | number,
		},
	},
	button: {
		height: '38px',
		font: {
			weight: 500 as string | number,
		},
		border: {
			width: '1px',
			radius: '20px',
			color: colors.text.main,
		},
		padding: {
			horizontal: '30px',
		},
	},
	radioButton: {
		border: {
			width: '1px',
			color: {
				dark: colors.quartenary.main,
				light: colors.quartenary.main,
			},
		},
		hover: {
			border: {
				color: {
					dark: colors.tertiary.main,
					light: colors.tertiary.main,
				},
			},
		},
		check: {
			color: {
				dark: 'white',
				light: 'white',
			},
		},
		// This actually sets the circle diameter to 6px;
		icon: { size: '12px' },
		gap: '10px',
		size: '20px',
	},
	select: {
		icons: {
			color: colors.secondary.main,
		},
		control: {
			extend: `color: ${colors.secondary.main}`,
		},
	},
	layer: {
		container: {
			zIndex: 40,
		},
		zIndex: 30,
	},
	text: {
		medium: {
			size: px(fontSizes[1]),
			height: lineHeight,
		},
	},
	tab: {
		extend: `padding: ${px(space[1])} ${px(space[3])}`,
		color: colors.secondary.main,
		margin: 'none',
		border: {
			size: 'xsmall',
			color: colors.quartenary.main,
			active: {
				color: colors.primary.main,
			},
			hover: {
				color: colors.quartenary.main,
			},
		},
		hover: {
			color: colors.primary.dark,
		},
		active: {
			color: colors.primary.main,
		},
	},
	tabs: {
		header: {
			extend: (props: any) => {
				// compact is passed from the Tabs component to all child
				// Tab components - but isn't directly available in 'props' :-(
				if (get(props, ['children', 0, 'props', 'compact'])) {
					return `
						flex-wrap: nowrap;
						overflow-x: auto;
					`;
				}
			},
		},
	},
	checkBox: {
		size: '20px',
		color: colors.primary.main,
		border: {
			color: {
				dark: colors.quartenary.main,
				light: colors.quartenary.main,
			},
			width: '1px',
		},
		check: {
			radius: '4px',
			thickness: '2px',
		},
		hover: {
			border: {
				color: {
					dark: colors.quartenary.main,
					light: colors.quartenary.main,
				},
			},
		},
		toggle: {
			color: {
				dark: colors.primary.main,
				light: colors.quartenary.main,
			},
			radius: '20px',
			size: '40px',
			knob: {},
		},
	},
	accordion: {
		border: {
			side: 'bottom',
		},
	},
};

export default theme;

export interface Theme extends Omit<typeof theme, 'colors'> {
	colors: Record<string, ColorVariants>;
}
